bot/cogs/moderation/incidents.py

```diff
@@ -0,0 +1,363 @@
+import asyncio
+import logging
+import typing as t
+from enum import Enum
+
+import discord
+from discord.ext.commands import Cog
+
+from bot.bot import Bot
+from bot.constants import Channels, Emojis, Roles, Webhooks
+from bot.utils.messages import sub_clyde
+
+log = logging.getLogger(__name__)
+
+# Amount of messages for `crawl_task` to process at most on start-up - limited to 50
+# as in practice, there should never be this many messages, and if there are,
+# something has likely gone very wrong
+CRAWL_LIMIT = 50
+
+# Seconds for `crawl_task` to sleep after adding reactions to a message
+CRAWL_SLEEP = 2
+
+
+class Signal(Enum):
+    """
+    Recognized incident status signals.
+
+    This binds emoji to actions. The bot will only react to emoji linked here.
+    All other signals are seen as invalid.
+    """
+
+    ACTIONED = Emojis.incident_actioned
+    NOT_ACTIONED = Emojis.incident_unactioned
+    INVESTIGATING = Emojis.incident_investigating
+
+
+# Reactions from roles not listed here will be removed
+ALLOWED_ROLES: t.Set[int] = {Roles.moderators, Roles.admins, Roles.owners}
+
+# Message must have all of these emoji to pass the `has_signals` check
+ALL_SIGNALS: t.Set[str] = {signal.value for signal in Signal}
+
+
+def make_username(reported_by: discord.Member, actioned_by: discord.Member, max_length: int = 80) -> str:
+    """
+    Create a webhook-friendly username from the names of `reported_by` and `actioned_by`.
+
+    If the resulting username length exceeds `max_length`, it will be capped at `max_length - 3`
+    and have 3 dots appended to the end. The default value is 80, which corresponds to the limit
+    Discord imposes on webhook username length.
+
+    If the value of `max_length` is < 3, ValueError is raised.
+    """
+    if max_length < 3:
+        raise ValueError(f"Maximum length cannot be less than 3: {max_length=}")
+
+    username = f"{reported_by.name} | {actioned_by.name}"
+    log.trace(f"Generated webhook username: {username} (length: {len(username)})")
+
+    if len(username) > max_length:
+        stop = max_length - 3
+        username = f"{username[:stop]}..."
+        log.trace(f"Username capped at {max_length=}: {username}")
+
+    return username
```

Comment:
So I added this to create the webhook username. It follows this format: `reported_by | actioned_by`. The format is easily adjustable so let me know if that doesn't work (using parens is an alternative).

![image](https://user-images.githubusercontent.com/44734341/85224830-4b3a4900-b3cd-11ea-917d-21cd710925b0.png)

Maximum webhook username length is 80 (request fails if above), so this will cut off anything above:

![image](https://user-images.githubusercontent.com/44734341/85224836-5b522880-b3cd-11ea-90c2-3fc1e0406956.png)

The thing is that both names and display names for regular users are capped at 32 so we should never actually exceed 80, but I did not realize that in time and I guess it doesn't hurt if the logic stays there in case something changes in the future. To trigger the cut-off for the screenshot I manually passed `max_len=40`.

Didn't use `textwrap.shorten` because I didn't want it to respect words / trim whitespace, I just wanted to cap the raw character count, which I believe is easier done manually.

How the `actioned_by` name is relayed was never really agreed upon with others, I failed to get a discussion started twice now, so I decided to just implement it and see how it works for everyone.

Reply:
Collapsing this as we scrapped this.